text/2019-03-19-calculated-commit-ts.md

```diff
@@ -0,0 +1,188 @@
+    is_ready: bool,
+
+    /// The version of the region's epoch. The `max_read_ts` can only be used
+    /// when `version` matches the prewrite request's version.
+    version: u64,
+
+    /// The maximum timestamp of transactional reads on this region. Note that
+    /// in order to ensure the `max_read_ts` is no less than reads on the region
+    /// before the region's leader goes to the current store, A timestamp will
+    /// be fetched from PD after creating or updating of the region.
+    max_read_ts: u64,
+}
+```
+
+`ReadTsCache` contains a map that maps from `region_id` to `RegionMaxTsRecord`:
+
+```rust
+// Divide the map into multiple according to region_id's hash to reduce lock
+// contention.
+max_read_ts_map: Vec<RwLock<HashMap<u64, Arc<RwLock<RegionMaxTsRecord>>>>>,
+```
+
+Each transactional reads need to update the `max_read_ts` of the requested
+region **before** getting snapshot (`u64::MAX` will be ignored). Each prewrite
+need to get the `max_read_ts` after successfully writing and return it with
+`PrewriteResponse`, and the TiKV client is responsible for calculating the
+`commit_ts` according to the timestamps carried in `PrewriteResponse`s. If
+failed to get the `max_read_ts` (due to not ready or version not match), just
+returns 0 to tell the client to fall back to allocating `commit_ts` from PD.
+
+`ReadTsCache` updates its inner region information via Raftstore observer and
+collects only leader. Every time a region is added to the map or a region's
+version is changed, `is_ready` will be unset until getting a timestamp from PD.
+
+### New constraints to TiKV's txn algorithm
+
+* If Txn1's `commit_ts` is the same as Txn2's `start_ts`, then Txn1's committed
+  data is visible to Txn2
+* If key `"k1"`'s latest version's `commit_ts` is the same as Txn1's `start_Ts`,
+  then Txn1 prewrites on key `"k1"` will cause write conflict.
+* In `CF_WRITE`, rollback record should never replace other records with the
+  same key and timestamp.
+
+### Code changes to TiKV Client
+
+In the 2PC algorithm, if all prewrite requests returns a non-zero timestamp,
+then the `commit_ts` can be calculated. Otherwise it should still request the
+`commit_ts` from PD.
+
+### Corner cases
+
+Regions' leaders are not static. They may transfer from a TiKV to another TiKV.
+Consider this case:
+
+1. Txn1 get start_ts 50
+2. Txn2 read `"k1"` with timestamp 200
+3. Leader transfer (And another TiKV doesn't know the previous `max_read_ts`
+   200)
+4. Txn3 read `"k1"` with timestamp 100
+5. Txn1 prewrite and commit on `"k1"`
+
+If we don't carefully handle this case, Txn1's `commit_ts` may be calculated
+101, then if Txn2 reads again, it will get a different value, which means, the
+Repeatable Read is broken.
+
+The similar problem also happens when merging occurs. If region1(`max_read_ts`
+is 200) merges into region2(`max_read_ts` is 100), then it's difficult to know
+that region2's `max_read_ts` should be updated to 200, especially if the two
```

Comment:
```suggestion
that whether  region2's `max_read_ts` should be updated to 200, especially when the leaders for the two
```

Reply:
I think both of these are ok:
* it's difficult to know that region2's `max_read_ts` should be updated to 200
* it's difficult to know that whether region2's `max_read_ts` should be updated
But it's strange to use both "whether" and "to 200"